### ISSUE TITLE
Give the code some structure and add abstractions

### ISSUE THREAD

**Issue by inf3rno** [noahbliss/mortar#4]
I identified two things in your code organization that I would call problems, but maybe they are ok for you current project size. I would just use more sophisticated methods, because it is a security related project.

1.)
Running `./0-initialsetup-prereqinstall.sh` is an unusual way to use a command line application. I would rather see `mortar --initial-setup` or something like that. Actually I think the best applications ask for user input instead of expecting it.

2.)
If you want to support multiple operating systems or multiple TPM versions, then you need abstractions and polymorphism at a certain level. In practice this means whenever you have multiple implementations of the same concept you come up with a general function name, something like `doStuff`. You move the different implementations of `doStuff` to different files and load them from there. So for example when you have tpm 1.2 then you import `tpm1.2/doStuff.sh` and when you have tpm 2.0 then you import `tpm2.0/doStuff.sh`. After you collected the configuration from the user in 1.) you can select the right implementations and import those files. After that you just call a general code, which calls functions like `doThisStuff()` without knowing the implementation details. A simple example of these can be the following:

`source run.sh`

run.sh
```sh
#!/usr/bin/env bash

run(){
	configure
	importFunctions
	main
}

configure(){
	isDebian=$(($RANDOM%2))
}

importFunctions(){
	if ((isDebian)); then
		source "./debian.sh"
	else
		source "./arch.sh"
	fi
	source "./main.sh"
}

run
```

main.sh
```sh
#!/usr/bin/env bash

main(){
	doStuff
	doOtherStuff
}
```

debian.sh
```sh
#!/usr/bin/env bash

doStuff(){
	echo "debian code running"
}

doOtherStuff(){
	echo "blah blah"
}
```

arch.sh
```sh
#!/usr/bin/env bash

doStuff(){
	echo "arch code running"
}

doOtherStuff(){
	echo "nooooooo yes"
}
```

As of the function names you can typically name them based on the comments you have in your code. For example here: https://github.com/noahbliss/mortar/blob/master/0-initialsetup-prereqinstall.sh#L14 you have `Figure out our distribuition. ` in comment and you can have a function with `figureOutOurDistribution` name instead. You can call that function from the `main`. If the `main` grows too complex then you group functions for example in your case `0-initialsetup-prereqinstall.sh` is the way of grouping. You just define `initialSetup` which calls the initial setup related functiions. You can keep them in a separate file and source it from `importFunctions` or from `main` if you don't want to scroll that much.

3.)
After you have 2.) you can write automated tests against the general code and mock out the OS or TPM version specific implementations. You will need a test runner instead of the application runner too. This way you can test maybe half of the code automagically. You can write integration tests for the rest if you want to, but I guess that would be hard even with virtualization.

`source test.sh`

test.sh
```sh
source "./mock.sh"
source "./main.sh"

testMain(){
	failures=0
	testIfDoStuffCalled
	testIfDoOtherStuffCalled
	if ((failures)); then
		echo "$failures tests failed"
	else
		echo "success"
	fi
}

testIfDoStuffCalled(){
	probeDoStuffCalled=0
	main
	if ((probeDoStuffCalled!=1)); then
		echo "doStuff should be called exactly once"
		((failures++))
	fi
}

testIfDoOtherStuffCalled(){
	probeDoOtherStuffCalled=0
	main
	if ((probeDoOtherStuffCalled!=1)); then
		echo "doOtherStuff should be called exactly once"
		((failures++))
	fi
}

testMain
```

mock.sh
```sh
#!/usr/bin/env bash

doStuff(){
	((probeDoStuffCalled++))
}

doOtherStuff(){
	((probeDoOtherStuffCalled++))
}
```

Not sure how much general code you will have that you can use with all OS types and all TPM versions, but if it is not just a few lines it is worth testing I think.

**Comment by noahbliss** [comment]
I gave this a once-over and sounds good. The code is absolutely due for a cleanup. Right now my primary focus is getting it functional on the primary distributions, but ultimately I'd like to take any distribution-specialized code out of the core "numbered" scripts and go from there. 

I also considered your point regarding doing away with the numbered scripts entirely. Ultimately that may be the best once the code is cleaned up and all the intended functionality is in place and would almost certainly be required if this were to be packaged into official repositories, but at the moment the separate files make troubleshooting easier and in my opinion, the installation a little more "idiot proof." 

The primary hang-up right now seems to be the need for multiple reboots which won't really ever go away. There are and should be hard "stops" a user needs to make when installing the framework. Their system may need some manual massaging in order to work properly (just yesterday I bought a motherboard that simply wouldn't work with secureboot enabled with custom keys). 

tl;dr good points. I'll stay the current design course for now, but will leave this issue open as a reminder. I greatly appreciate the input!

**Comment by inf3rno** [comment]
@noahbliss 

> The primary hang-up right now seems to be the need for multiple reboots which won't really ever go away. There are and should be hard "stops" a user needs to make when installing the framework. Their system may need some manual massaging in order to work properly (just yesterday I bought a motherboard that simply wouldn't work with secureboot enabled with custom keys).

I guess it is possible to save the configuration, the actual state and the previous state and register mortar to the init system. So after reboot we could start mortar automatically and offer the user to continue or rollback the changes we made so far. Making it user friendly at this level requires a lot more work though.

In my experience these "cleanups" almost never happen, because the code "works" without them and they feel like wasted time. I like to write programs this way from the beginning, because it is a lot easier to develop and maintain a properly organized code. I do automated testing to make sure that changes don't break features implemented previously. In this case it might not be necessary, at least I don't think we can do tests automatically without real hardware. At least I don't know what percent we have hardware specific and OS specific code.

I think I'll buy a TPM for my home server and experiment with this too.

**Comment by inf3rno** [comment]
@noahbliss I found many emulators too meanwhile. :-) Here you got a collection: https://github.com/PeterHuewe There are emulators for 1.2 and 2.0 too. Someone even created a docker image based on these emulators: https://github.com/starlab-io/docker-tpm-emulator I think it is possible to use that for testing or maybe write our own docker images because this one is for Ubuntu only. Some of the emulators work on many different OS-es including BSD. There is even an emulator that runs in the browser: https://google.github.io/tpm-js/ I am not sure, maybe that would be too low level to us. I guess we just need to mock out some of the high level TPM dependencies like trousers or installing them for the emulator somehow. In theory it is possible: https://stackoverflow.com/questions/30052769/running-marios-tpm-emulator-with-tss-test-suit So you can write tests for docker or VM on the emulated TPM without burning real TPMs. I am not sure how much I can contribute. I am not into bash scripting that much and I need to work on my own hobby projects.

**Comment by noahbliss** [comment]
Structure added with distro-specific code removed from main scripts. There is more to go, but this is tangible gains towards code cleanliness. I'll close this issue, but feel free to submit another for specific improvements/etc. Thanks!

**Comment by inf3rno** [comment]
@noahbliss Looks better. You left a few comments in https://github.com/noahbliss/mortar/blob/master/0-initialsetup-prereqinstall.sh

**Comment by noahbliss** [comment]
Thanks! Resolved.
